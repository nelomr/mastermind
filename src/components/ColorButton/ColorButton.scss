.color-button {
  width: 20px;
  height: 20px;
  color: $red-color;
  cursor: pointer;
  background: transparent;
  border-radius: 50%;
  border: 1px solid transparent;

  @include breakpoint($min-desktop) {
    width: 40px;
    height: 40px;
  }

  &.is-white {
    border-color: $black-color;
  }

  &.is-option-color {
    width: 40px;
    height: 40px;
    margin-right: 20px;
    margin-bottom: 0;
    opacity: 0;
    box-shadow: 0 0 0 0 currentColor;

    &:last-child {
      margin-right: 0;
      margin-bottom: 0;
    }
  
    .is-loaded & {
      opacity: 1;
      transition: all .2s ease;
  
      @for $i from 1 through 8 {
        &:nth-child(#{$i}) {
          transition-delay: ($i * 0.1s);
        }
      }
    }

    &.is-white {
      border-color: transparent;
    }
    
  
    &.is-selected {
      box-shadow: 0 0 12px 2px currentColor;

      &.is-white {
        box-shadow: 0 0 12px 2px darkgray;
      }
    }
  
    @include breakpoint($min-desktop) {
      margin-right: 0;
      margin-bottom: 20px;
    }
  }
}

