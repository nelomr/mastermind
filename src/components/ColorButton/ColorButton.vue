<template>
  <div 
    @click="manageEvent(index)"
    class="color-button"
    :class="{'is-option-color': !isGuessButton, 'is-white': colorDefined === 'white'}"
    :style="`background-color: ${ colorDefined }; color: ${ colorDefined };`"
  >
  </div>
</template>

<script>
import EventBus from '@/modules/event-bus';
import { mapMutations, mapState, mapActions } from 'vuex';

export default {
  name: 'ColorButton',
  props: {
    color: {
      type: String,
      default: 'lightgray'
    },
    index: {
      type: Number,
      default: 0
    },
    indexRow: {
      type: Number,
      default: 0
    },
    isGuessButton: {
      type: Boolean,
      default: false
    }
  },
  data() {
    return {
      colorDefined: this.color,
    }
  },
  computed: {
    ...mapState([
      'currentColor',
      'guessRowColors',
      'row'
    ])
  },
  created() {
    EventBus.$on('reset', () => {
      this.reset();
    });
    EventBus.$on('resetRow', () => {
      this.resetRow();
    });
  },
  methods: {
    ...mapMutations([
      'setCurrentColor'
    ]),
    ...mapActions([
      'addColorToCheck'
    ]),
    manageEvent(index) {
      this.isGuessButton ? this.setColor(index) : this.getColor(index);
    },
    getColor(index) {
      this.setCurrentColor(this.color);
      this.$emit('clicked', index);
    },
    setColor(index) {
      this.currentColor != null && this.addColorToCheck({color: this.currentColor, indexArray: index});
      this.colorDefined = this.currentColor;
    },
    reset() {
      Object.assign(this.$data,this.$options.data.call(this));
    },
    resetRow() {
      (this.row === this.indexRow) && Object.assign(this.$data,this.$options.data.call(this));
    }
  }
}
</script>
<style lang="scss">
  @import "./ColorButton.scss";
</style>