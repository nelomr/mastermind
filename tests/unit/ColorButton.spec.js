import Vuex from 'vuex';
import { shallowMount, createLocalVue } from '@vue/test-utils'
import ColorButton from '@/components/ColorButton/ColorButton.vue'

const localVue = createLocalVue();
localVue.use(Vuex);

describe('ColorButton.vue', () => {
  let color = 'lightgray';
  let index = 0;
  let indexRow = 0;
  let isGuessButton = false;
  let row = 0;
  let status = 'running';
  let currentColor = 'blue';
  
  let props;
  let computed;
  let actions;
  let store;

  beforeEach(() => {
    props = {
      color,
      index,
      indexRow,
      isGuessButton
    };
    computed = {
      row() {
          return row;
      },
      status() {
          return status;
      },
      currentColor() {
        return currentColor;
      }
    };
    actions = {
      addColorToCheck: jest.fn(),
      setCurrentColor: jest.fn()
    };
    store = new Vuex.Store({
      actions
    });
  });

  const build = () => {
    const wrapper = shallowMount(ColorButton, {
      propsData: props,
      computed,
      mocks: {$store: store}
    });

    return {
      wrapper,
      button: () => wrapper.find('.color-button')
    };
  };

  it('renders the component', () => {
    const {wrapper} = build();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('color default on init', () => {
    const {wrapper} = build();
   
    expect(wrapper.vm.colorDefined).toEqual('lightgray');
  });

  it('renders the option button component', () => {
    const {wrapper} = build();

    expect(wrapper.find('.color-button').classes('is-option-color')).toBe(true);
  });

  it('renders the guess button component', async() => {
    const {wrapper} = build();

    wrapper.setProps({
      isGuessButton: true,
    });
    await wrapper.vm.$nextTick();

    expect(wrapper.find('.color-button').classes('is-option-color')).toBe(false);
  });

  it('get color clicked on button option component', () => {
    const {wrapper, button} = build();
    const getColor = jest.spyOn(wrapper.vm, 'getColor'); 

    button().trigger('click');
    expect(getColor).toHaveBeenCalled();
    expect(actions.setCurrentColor).toHaveBeenCalled();
  });

  it('set color clicked on button guess component', () => {
    const {wrapper, button} = build();
    const setColor = jest.spyOn(wrapper.vm, 'setColor'); 

    wrapper.setProps({
      isGuessButton: true,
    });

    button().trigger('click');
    expect(setColor).toHaveBeenCalled();
    expect(actions.addColorToCheck).toHaveBeenCalled();
  });
})
