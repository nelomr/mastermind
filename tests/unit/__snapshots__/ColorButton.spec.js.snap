// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorButton.vue renders the component 1`] = `<div class="color-button is-option-color" style="background-color: lightgray; color: lightgray;"></div>`;
